### PR TITLE
Add support for java11 in gae

### LINKED ISSUE #10196: Add Java11 support to Google App Engine Generator
##### **Overview of the feature request**

Currently the GAE generator can generate deployment files for java 8 (https://cloud.google.com/appengine/docs/standard/java/runtime). However starting recently GAE supports java 11 (https://cloud.google.com/appengine/docs/standard/java/runtime). I suggest we develop a branch of our GAE generator to support java 11. My idea is to initially give the user a choice of choosing between java  8 and 11 and generate the files based on the choice. 

On another note it is possible to run our jhipster project on java 11 on GAE standard as I've done on [>>this<<](https://github.com/python-sinhala-education-society/OpenLearnr) project. The only things we need is to remove the ```appengine-web.xml``` and add a ```app.yml```. Complete migration summary is given here; https://cloud.google.com/appengine/docs/standard/java11/java-differences

If this is something useful, I can help with doing the PR for this. Do we need to fix the current GAE generator before attempting this? 

##### **Motivation for or Use Case**

jhipster supports java 11 and recently we migrated the Travis build process to java 11 as well (https://github.com/jhipster/generator-jhipster/pull/10175), it seems obvious that we need to add support for java 11 GAE generators at some point. 

##### **Related issues or PR**

#10175, #9605 

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by SudharakaP**
This migrates our existing GAE generator from Java 8 to Java 11. Note that this will remove all support for Java 8 GAE config file generation (in other words after this fix is merged users will only be able to generate Java 11 GAE configs). I thought this is the best since giving users the ability to choose between Java 8 and Java 11 seems messy since we have migrated other parts of jhipster-generator (such as the travis build to java 11) completely. Also the configs and the way GAE deployments work is very different from Java 8 to Java 11; and Java 11 uses JAR deployment which simplifies matters a lot.

Fix #10196

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/gae/index.js [hunk 1]
```diff
@@ -363,17 +366,41 @@ module.exports = class extends BaseGenerator {
...
                 if (this.prodDatabaseType !== 'mysql' && this.prodDatabaseType !== 'mariadb' && this.prodDatabaseType !== 'postgresql')
                     return;
 
                 const done = this.async();
 
                 const cloudSqlInstances = [{ value: '', name: 'New Cloud SQL Instance' }];
-
                 exec(
-                    `gcloud sql instances list  --format='value[separator=":"](project,region,name)' --project="${this.gcpProjectId}"`,
+                    `gcloud sql instances list  --format="value[separator=":"](project,region,name)" --project="${this.gcpProjectId}"`,
```
Comment (PierreBesson) [inline generators/gae/index.js]: I'm pretty sure there is a mistake here with the `"`

File: generators/gae/templates/app.yaml.ejs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+<%_ if (gaeScalingType === 'manual') { _%>
+manual_scaling:
+    instances: <%= gaeInstances %>
+<%_ } else if (gaeScalingType === 'basic') { _%>
+basic_scaling:
+    <%_ if (gaeMaxInstances > 0) { _%>
+    max_instances: <%= gaeMaxInstances %>
+    <%_ } _%>
+<%_ } else if (gaeScalingType === 'automatic') { _%>
+automatic_scaling:
```
Comment (PierreBesson) [inline generators/gae/templates/app.yaml.ejs]: I have an error when answering the default values in the prompt of `min_instance: 0` and `max_instance: 0`, in this cas the `automatic_scaling:` block should disappear entirely. https://cloud.google.com/appengine/docs/flexible/java/reference/app-yaml#automatic_scaling

File: generators/gae/templates/app.yaml.ejs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+runtime: java11
```
Comment (PierreBesson) [inline generators/gae/templates/app.yaml.ejs]: Maybe start this file with the following:
```
# ========================================================================
# App Engine Configuration.
# Full reference is available at:
# https://cloud.google.com/appengine/docs/flexible/java/reference/app-yaml
# ========================================================================

**Comment by PierreBesson** [comment]
@SudharakaP For some reason when running `jhipster gae`, I have an error when generating `application-prod-gae.yml`

```
SyntaxError: Nested mappings are not allowed in compact mappings (18:19)
  16 |         url: jdbc:mysql://google/gaedemo?cloudSqlInstance=proven-serenity-238515:northamerica-northeast1:demo1
  17 | &socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false
> 18 |         username: root
     |                   ^^^^^^^^^^^^^
> 19 |         password: 
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 20 |         hikari:
     | ^
  21 |             maximumPoolSize: 8
  22 | 
    at e (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:261)
    at Object.parse (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:135253)
    at Object.parse$2 [as parse] (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:10639:19)
    at coreFormat (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:13856:23)
    at format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14115:73)
    at formatWithCursor (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14131:12)
    at /home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42399:15
    at Object.format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42418:12)
    at prettier.resolveConfig.then.options (/home/pierre/github/generator-jhipster/generators/generator-transforms.js:44:39)
```
Can you add quotes in the application-prod-gae.yml template aound values like this:
```
spring:
    datasource:
        type: com.zaxxer.hikari.HikariDataSource
        url: "jdbc:mysql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false"
        username: <%=gcpCloudSqlUserName%>
        password: <%=gcpCloudSqlPassword%>
        hikari:
            maximumPoolSize: 8
    datasource:
        type: com.zaxxer.hikari.HikariDataSource
        url: "jdbc:postgresql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.postgres.SocketFactory"
        username: "<%=gcpCloudSqlUserName%>"
        password: "<%=gcpCloudSqlPassword%>"
```

**Comment by saturnism** [comment]
@SudharakaP this is awesome! thank you.

I didn't see app engine plugin version being updated. It may need bump. https://cloud.google.com/appengine/docs/standard/java11/using-maven#adding_the_maven_and_eclipse_jetty_maven_plugins

/cc @ludoch

**Comment by SudharakaP** [comment]
@saturnism : You are welcome. I'll check and fix all the things mentioned by Pierre soon. Then hopefully we'll have a working app engine generator. Although I do have some more plans for this one. Will create some more prs in the future. :smile:

File: generators/gae/index.js (hunk 1 above)
Comment (SudharakaP) [reply generators/gae/index.js]: Thanks. Corrected 😄

File: generators/gae/templates/app.yaml.ejs (hunk 2 above)
Comment (SudharakaP) [reply generators/gae/templates/app.yaml.ejs]: I've made changes to this. I think one should be allowed to choose max_instances: 0 and/or min instances: 0. These are allowed settings in GCP and I don't think we need the block to disappear. If max_instances: 0 is chosen the setting will be disabled anyways. I've made an error in my previous commit such that when 0 is chosen the max/min_instances blocks doesn't appear. 

Let me know your thoughts; I also see that the reference doc you pointed seems to the flexible environment not the standard one which is what we support; the correct doc is; https://cloud.google.com/appengine/docs/standard/java11/config/appref#automatic_scaling
The way standard handles max/min_instances is a bit different. 

Supporting flexible is a different discussion which I'll bring up in perhaps another PR. 😄

**Comment by SudharakaP** [comment]
> @SudharakaP For some reason when running `jhipster gae`, I have an error when generating `application-prod-gae.yml`
> 
> ```
> SyntaxError: Nested mappings are not allowed in compact mappings (18:19)
>   16 |         url: jdbc:mysql://google/gaedemo?cloudSqlInstance=proven-serenity-238515:northamerica-northeast1:demo1
>   17 | &socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false
> > 18 |         username: root
>      |                   ^^^^^^^^^^^^^
> > 19 |         password: 
>      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > 20 |         hikari:
>      | ^
>   21 |             maximumPoolSize: 8
>   22 | 
>     at e (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:261)
>     at Object.parse (/home/pierre/github/generator-jhipster/node_modules/prettier/parser-yaml.js:1:135253)
>     at Object.parse$2 [as parse] (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:10639:19)
>     at coreFormat (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:13856:23)
>     at format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14115:73)
>     at formatWithCursor (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:14131:12)
>     at /home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42399:15
>     at Object.format (/home/pierre/github/generator-jhipster/node_modules/prettier/index.js:42418:12)
>     at prettier.resolveConfig.then.options (/home/pierre/github/generator-jhipster/generators/generator-transforms.js:44:39)
> ```
> 
> Can you add quotes in the application-prod-gae.yml template aound values like this:
> 
> ```
> spring:
>     datasource:
>         type: com.zaxxer.hikari.HikariDataSource
>         url: "jdbc:mysql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.mysql.SocketFactory&useSSL=false"
>         username: <%=gcpCloudSqlUserName%>
>         password: <%=gcpCloudSqlPassword%>
>         hikari:
>             maximumPoolSize: 8
>     datasource:
>         type: com.zaxxer.hikari.HikariDataSource
>         url: "jdbc:postgresql://google/<%=gcpCloudSqlDatabaseName%>?cloudSqlInstance=<%=gcpCloudSqlInstanceName%>&socketFactory=com.google.cloud.sql.postgres.SocketFactory"
>         username: "<%=gcpCloudSqlUserName%>"
>         password: "<%=gcpCloudSqlPassword%>"
> ```

@PierreBesson : Thanks for all the code reviews; I've fixed them all in my latest commit. Please let me know if you see any further issues. 😄 

> @SudharakaP this is awesome! thank you.
> 
> I didn't see app engine plugin version being updated. It may need bump. https://cloud.google.com/appengine/docs/standard/java11/using-maven#adding_the_maven_and_eclipse_jetty_maven_plugins
> 
> /cc @ludoch

@saturnism : Thanks for pointing this out. I've updated the plugin version as well to 2.1.0. 😄

File: generators/gae/templates/app.yaml.ejs (hunk 3 above)
Comment (SudharakaP) [reply generators/gae/templates/app.yaml.ejs]: Thanks. Done 😄

File: generators/gae/templates/app.yaml.ejs (hunk 2 above)
Comment (PierreBesson) [reply generators/gae/templates/app.yaml.ejs]: Yes you are correct on both points.

**Comment by PierreBesson** [comment]
Thank you @SudharakaP for the hard work. Very appreciated. Looking forward to review your next contributions on app engine support. BTW you can remove the `BETA` log at startup as it doesn't do justice to this subgen.

**Comment by SudharakaP** [comment]
@PierreBesson : You are welcome; it's fun contributing to jhipster. 

Btw, I would suggest keeping the ```BETA``` log for a week or so until I get some other corner cases right. I think there's still issues which could cause confusion in this sub-generator. I've created a new issue for discussion; https://github.com/jhipster/generator-jhipster/issues/10331. 

cc: @saturnism 

Please feel free to add any suggestions. :smile:

**Comment by jdubois** [comment]
@SudharakaP thanks a lot for all your work and contributions!!! I also saw you sponsored the project ( https://opencollective.com/sudharaka-palamakumbura )!! Do you have a Twitter handle, so I can thank you publicly?

**Comment by SudharakaP** [comment]
@jdubois: You are welcome, and hope it makes a difference. I don't use Twitter :smile:

**Pull request merged**